ykcapi/src/lib.rs

```diff
@@ -48,6 +52,122 @@ pub extern "C" fn __ykutil_get_llvmbc_section(res_addr: *mut *const c_void, res_
     }
 }
 
+/// Reads out registers spilled to the stack of the previous frame during the deoptimisation
+/// routine. The order of the registers are in accordance to the DWARF register number mapping
+/// referenced in the SystemV ABI manual (https://uclibc.org/docs/psABI-x86_64.pdf).
+struct Registers {
+    addr: *const usize,
+}
+
+impl Registers {
+    /// Creates a Registers struct from a given a pointer on the stack containing spilled
+    /// registers.
+    fn from_ptr(ptr: *const c_void) -> Registers {
+        Registers {
+            addr: ptr as *const usize,
+        }
+    }
+
+    /// Read the spilled register value at the offset `off` from the previous stack frame.
+    fn read_from_stack(&self, off: isize) -> usize {
+        unsafe { ptr::read::<usize>(self.addr.offset(off)) }
+    }
+
+    /// Retrieve the previous frame's register value given by its DWARF register number `id`. This
+    /// number additionally functions as an offset into the the spilled stack to find that
+    /// register's value.
+    #[cfg(target_arch = "x86_64")]
+    fn get(&self, id: u16) -> usize {
```

Comment:
Either this function or `from_ptr` should be unsafe as the address field could be a dangling or otherwise invalid pointer.

Reply:
Thanks! Fixed in 04406d1